src/edu/washington/escience/myria/operator/CountFilter.java

```diff
@@ -0,0 +1,214 @@
+
+  /**
+   * The logger for this class.
+   * */
+  static final Logger LOGGER = LoggerFactory.getLogger(CountFilter.class);
+
+  /**
+   * Indices to unique tuples.
+   * */
+  private transient IntObjectHashMap<IntArrayList> uniqueTupleIndices;
+
+  /**
+   * The buffer for storing unique tuples.
+   * */
+  private transient MutableTupleBuffer uniqueTuples = null;
+  /**
+   * The count of each unique tuple.
+   * */
+  private transient MutableTupleBuffer tupleCounts = null;
+
+  /** threshold of the count. */
+  public final int threshold;
+
+  /** column indices of the key. */
+  private final int[] keyColIndices;
+
+  /**
+   * @param threshold threshold
+   * @param keyColIndices key column indices.
+   */
+  public CountFilter(final int threshold, final int[] keyColIndices) {
+    Preconditions.checkArgument(threshold >= 0, "threshold needs to be greater than or equal to 0");
+    this.threshold = threshold;
+    this.keyColIndices = Arrays.copyOf(keyColIndices, keyColIndices.length);
+  }
+
+  @Override
+  public void cleanup() {
+    uniqueTuples = null;
+    uniqueTupleIndices = null;
+    tupleCounts = null;
+  }
+
+  /**
+   * Do duplicate elimination for tb.
+   * 
+   * @param tb the TupleBatch for performing DupElim.
+   * @return the duplicate eliminated TB.
+   * */
+  protected TupleBatch countFilter(final TupleBatch tb) {
+    final int numTuples = tb.numTuples();
+    if (numTuples <= 0) {
+      return tb;
+    }
+    doCount.inputTB = tb;
+    final List<? extends Column<?>> columns = tb.getDataColumns();
+    final BitSet toRemove = new BitSet(numTuples);
+    for (int i = 0; i < numTuples; ++i) {
+      final int nextIndex = uniqueTuples.numTuples();
+      final int cntHashCode = HashUtils.hashSubRow(tb, keyColIndices, i);
+      IntArrayList tupleIndexList = uniqueTupleIndices.get(cntHashCode);
+      if (tupleIndexList == null) {
+        tupleIndexList = new IntArrayList();
+        uniqueTupleIndices.put(cntHashCode, tupleIndexList);
+      }
+      doCount.found = false;
+      doCount.meet = false;
+      doCount.sourceRow = i;
+      tupleIndexList.forEach(doCount);
+      if (!doCount.found) {
+        for (int j : keyColIndices) {
+          uniqueTuples.put(j, columns.get(j), i);
+        }
+        tupleCounts.put(0, new IntArrayColumn(new int[] { 1 }, 1), 0);
+        tupleIndexList.add(nextIndex);
+        if (threshold <= 1) {
+          doCount.meet = true;
+        }
+      }
+      if (!doCount.meet) {
+        toRemove.set(i);
+      }
+    }
+    return tb.filterOut(toRemove).selectColumns(keyColIndices);
+  }
+
+  @Override
+  public Schema getSchema() {
+    Schema schema = getOp().getInputSchema();
+    if (schema == null) {
+      return null;
+    }
+    return schema.getSubSchema(keyColIndices);
+  }
+
+  @Override
+  public void init(final ImmutableMap<String, Object> execEnvVars) {
+    uniqueTupleIndices = new IntObjectHashMap<IntArrayList>();
```

Comment:
Eh, I see you're only storing 4 bytes of the 16-byte hash code. This could be upgraded to 8 bytes (`LongIntMap`), which should be enough to prevent collisions before about `2^32` entries. Does that seem safe enough? Even if we're doubling the key size, wouldn't this still be lower-overhead than storing an `IntArrayList` for every hash code? I guess we could use `BigInteger` as a hash key, but that seems like it would be very high-overhead and defeat the whole optimization...